plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/core/explorer/webview/Webview.kt

```diff
@@ -53,57 +69,207 @@ private class WebviewDialog(private val project: Project) : DialogWrapper(projec
 }
 
+class ToolkitWebviewPanel(val project: Project) {
+    private val webviewContainer = Wrapper()
+    var browser: ToolkitWebviewBrowser? = null
+        private set
+
+    val component = panel {
+        row {
+            cell(webviewContainer)
+                .horizontalAlign(HorizontalAlign.FILL)
+                .verticalAlign(VerticalAlign.FILL)
+        }.resizableRow()
+
+        row {
+            cell(
+                JButton("Show Web Debugger").apply {
+                    addActionListener(
+                        ActionListener {
+                            browser?.jcefBrowser?.openDevtools()
+                        },
+                    )
+                },
+            )
+                .horizontalAlign(HorizontalAlign.CENTER)
+                .verticalAlign(VerticalAlign.BOTTOM)
+        }
+    }
+
+    init {
+        if (!JBCefApp.isSupported()) {
+            // Fallback to an alternative browser-less solution
+            webviewContainer.add(JBTextArea("JCEF not supported"))
+            browser = null
+        } else {
+            browser = ToolkitWebviewBrowser(project).also {
+                webviewContainer.add(it.component())
+            }
+        }
+    }
+
+    companion object {
+        fun getInstance(project: Project?) = project?.service<ToolkitWebviewPanel>() ?: error("")
+    }
+}
+
 // TODO: STILL WIP thus duplicate code / pending move to plugins/toolkit
-class ToolkitWebviewBrowser(val project: Project) {
-    val jcefBrowser: JBCefBrowserBase by lazy {
+class ToolkitWebviewBrowser(val project: Project) : LoginBrowser(project, ToolkitWebviewBrowser.DOMAIN) {
+    override val jcefBrowser: JBCefBrowserBase by lazy {
         val client = JBCefApp.getInstance().createClient()
-
         Disposer.register(project, client)
-
         JBCefBrowserBuilder()
             .setClient(client)
             .setOffScreenRendering(true)
+            .setCreateImmediately(true)
             .build()
     }
+    override val query: JBCefJSQuery = JBCefJSQuery.create(jcefBrowser)
+
+    override val handler = Function<String, JBCefJSQuery.Response> {
+        val command = jacksonObjectMapper().readTree(it).get("command").asText()
+        getLogger<ToolkitWebviewBrowser>().debug { "Command received from the browser: $command" }
+
+        when (command) {
+            "prepareUi" -> {
+                this.prepareBrowser(BrowserState(FeatureId.AwsExplorer))
+            }
+
+            "loginBuilderId" -> {
+                runInEdt {
+                    Login.BuilderId(CODECATALYST_SCOPES, onPendingAwsId).loginBuilderId(project)
+                }
+            }
+
+            "loginIdC" -> {
+                val profileName = jacksonObjectMapper().readTree(it).get("profileName").asText()
+                val url = jacksonObjectMapper().readTree(it).get("url").asText()
+                val region = jacksonObjectMapper().readTree(it).get("region").asText()
```

Comment:
definitely feels like we're still duplicating things

Reply:
yea lots of cleanup work pending.

Reply:
and also, it's not type safe as you said